File: kotlin/hello_world.kt
fun insertSort(arr: Array<Int>) {
    for (i in 1 until arr.size) {
        //println("loop i = $i")
        for (j in (i - 1) downTo 0) {
            //println("loop j = $j")
            if (arr[j] > arr[j + 1]) {
                // swap element j and j + 1
                arr[j] = arr[j + 1].also {arr[j + 1] = arr[j]}
            } else {
                //println("break, i = $i, j = $j")
                break
            }
        }
    }
}

fun printArray(arr: Array<Int>) {
    print("(")
    for (i in arr) {
        print("$i, ")
    }
    println(")")
}

fun main() {
    println("Hello, World!")
    val a = arrayOf(45, 34, 78, 12, 34, 32, 29, 64)
    val b = arrayOf(12, 29, 32, 34, 34, 45, 64, 78)
    insertSort(a)

    printArray(a)
    for (i in 0 until a.size) {
        if (a[i] != b[i]) {
            println("ERROR: NOT Equals!!! a[$i](${a[i]}) != b[$i](${b[i]})")
            return
        }
    }

    println("Equals!!!")
}

File: go/hello_world.go
package main

import "fmt"

/**
  * 插入排序，移动插入版
  */
func insertSort(arr []int, n int) []int {
    for i := 1; i < n; i++ {
        tmp := arr[i]
        j := i - 1
        for ; j >= 0; j-- {
            if arr[j] > tmp {
                arr[j + 1] = arr[j]
            } else {
                break
            }
        }
        arr[j + 1] = tmp
    }
    return arr
}

func printArray(arr []int, n int) {
    for i := 0; i < n; i++ {
        fmt.Println("arr: ", i, arr[i])
    }
}

func assertArrayEquals(arr []int, expect []int, n int) bool {
    for i := 0; i < n; i++ {
        if expect[i] != arr[i] {
            return false
        }
    }
    return true
}
/**
  * tasks:
  * declare an array to be sorted fill with some int
  * declare a method, passed the array as its input param
  * assert the sorted array by for loop iterator
  */
func main() {
    fmt.Println("Hello, World!")

    fmt.Println("Insert Sort begin")
    a := []int{45, 34, 78, 12, 34, 32, 29, 64}
    b := []int{12, 29, 32, 34, 34, 45, 64, 78}
    //printArray(a, len(a))
    insertSort(a, len(a))
    //printArray(a, len(a))
    //printArray(b, len(b))
    if assertArrayEquals(a, b, len(a)) {
        fmt.Println("Equals!!!")
    } else {
        fmt.Println("NOT Equals!!!")
    }
    fmt.Println("Insert Sort end")
}
